File: check-if-binary-string-has-at-most-one-segment-of-ones/solution1784.py
# https://leetcode.com/problems/check-if-binary-string-has-at-most-one-segment-of-ones/

class Solution1784:

    # T: O(n), S: O(1)
    def checkOnesSegment(self, s: str) -> bool:
        n = len(s)
        for i in range(n - 1):
            curr, next = s[i], s[i + 1]
            if curr == '0' and next == '1':
                return False

        return True

    def checkOnesSegmentBuiltin(self, s: str) -> bool:
        return '01' not in s


def main() -> None:
    sol = Solution1784()
    print(sol.checkOnesSegment("1001"))  # False
    print(sol.checkOnesSegmentBuiltin("1001"))  # False
    print(sol.checkOnesSegment("110"))  # True
    print(sol.checkOnesSegment("10"))  # True


if __name__ == "__main__":
    main()

File: check-if-binary-string-has-at-most-one-segment-of-ones/Solution1784.java
// https://leetcode.com/problems/check-if-binary-string-has-at-most-one-segment-of-ones/

class Solution1784 {

    // T: O(n), S: O(1)
    public boolean checkOnesSegment(String s) {
        // If pattern 101 exists, return false
        for (int i = 0; i < s.length() - 1; i++) {
            char curr = s.charAt(i);
            char next = s.charAt(i + 1);
            if (curr == '0' && next == '1') {
                return false;
            }
        }
        return true;
    }

    public boolean checkOnesSegmentBuiltin(String s) {
        return !s.contains("01");
    }

    public static void main(String[] args) {
        var sol = new Solution1784();
        System.out.println(sol.checkOnesSegment("1001")); // false
        System.out.println(sol.checkOnesSegmentBuiltin("1001")); // false
        System.out.println(sol.checkOnesSegment("110")); // true
        System.out.println(sol.checkOnesSegment("10")); // true
    }

}
